Data Layer: Asynchronizerize actual data layer handlers

dmsnell opened the pull request:
See #25073

This patch breaks apart the data-layer handlers from the initial boot of Calypso.
It adds the necessary imports to make sure that `webpack` includes the associated
data layer handlers whenever the first time is that an action of the given type could
possibly be dispatched. This preserves the synchronous and instant operation of the
middleware while allowing for our handlers to be split into different chunks.

**Testing**

Smoke test Calypso all over the place. Be sure to manually test important user flows
like signup, checklists, Happychat sessions, creating a post, subscribing and unsubscribing
to different feeds, etc…

**Status**

Rebased but had to manually merge in the changes from https://github.com/Automattic/wp-calypso/pull/27245 and from https://github.com/Automattic/wp-calypso/pull/27090.
@adamziel and @michaeldcain I'd be grateful if you could make sure this
doesn't break what you changed.

**Todo**
 - [x] We need to get these last bits in though they don't have to be pulled in with this PR.
 - [ ] Need to update the data layer documentation to make sure we don't start bloating `build` again

**Size**
http://iscalypsofastyet.com/branch?branch=data-layer/asynchronizerize-concierge
```
Delta:
chunk                                                                                                          stat_size             parsed_size             gzip_size
accept-invite                                                                                                    -1797 B    (-3.2%)      -1193 B    (-3.7%)     -264 B    (-3.6%)
accept-invite~checkout~jetpack-connect~signup                                                                    -1825 B    (-5.3%)      -1422 B    (-8.5%)     -480 B    (-9.8%)
accept-invite~jetpack-connect~signup                                                                             +2171 B    (+3.0%)       +570 B    (+1.4%)     +205 B    (+2.1%)
account                                                                                                          +2179 B    (+4.5%)      +1486 B    (+5.0%)     +297 B    (+3.8%)
account-close                                                                                                    +1753 B    (+4.3%)       +790 B    (+3.6%)     +272 B    (+4.7%)
account-recovery                                                                                                 +8510 B   (+13.4%)      +3438 B   (+12.3%)     +688 B    (+9.6%)
account~account-close~happychat~help~me~notification-settings~privacy~purchases~security                        +16750 B   (+12.7%)      +4252 B    (+5.7%)     +966 B    (+5.6%)
account~ads~async-load-design~async-load-design-blocks~async-load-design-playground~async-load-guten~4da1d48e       +0 B                    +0 B                  -1 B    (-0.0%)
account~async-load-design~async-load-design-playground~async-load-gutenberg-blocks~async-load-gutenb~8b5583a0   -31087 B  (deleted)     -17021 B  (deleted)    -4581 B  (deleted)
activity~signup~stats                                                                                           +39113 B   (+64.3%)     +17421 B   (+56.1%)    +3699 B   (+49.0%)
activity~stats                                                                                                   -5478 B    (-2.7%)      -3643 B    (-3.7%)     -580 B    (-2.5%)
ads                                                                                                              +2825 B    (+3.4%)      +1088 B    (+2.2%)     +256 B    (+2.3%)
async-load-blocks-calendar-popover~comments~post-editor                                                             +0 B                    +0 B                  -1 B    (-0.0%)
async-load-blocks-inline-help-popover~help                                                                       +4134 B    (+2.1%)      +1539 B    (+1.3%)     +400 B    (+1.5%)
async-load-components-happychat                                                                                     +0 B                    +0 B                  -1 B    (-0.1%)
async-load-components-web-preview-component                                                                      -5184 B    (-5.5%)      -3121 B    (-6.2%)     -618 B    (-5.4%)
async-load-components-web-preview-component~posts-custom~posts-pages                                             +3861 B   (+12.8%)      +2337 B   (+14.1%)     +409 B    (+9.7%)
async-load-components-web-preview-component~reader                                                              +38144 B   (+13.6%)     +12138 B    (+8.3%)     -935 B    (-2.7%)
async-load-design                                                                                                +3622 B    (+0.8%)      +2615 B    (+0.9%)     +361 B    (+0.7%)
async-load-design-blocks                                                                                       +220288 B   (+12.6%)    +111511 B   (+12.3%)   +26289 B   (+13.5%)
async-load-design-playground                                                                                     +3622 B    (+0.9%)      +2615 B    (+1.0%)     +446 B    (+0.9%)
async-load-design~async-load-design-blocks~async-load-design-playground~async-load-gutenberg-blocks~~0742ab94    -1797 B    (-5.1%)      -1193 B    (-6.0%)     -277 B    (-5.5%)
async-load-design~async-load-design-blocks~async-load-design-playground~async-load-gutenberg-blocks~~31e5b9f5   -38553 B  (deleted)     -33933 B  (deleted)   -10319 B  (deleted)
async-load-design~async-load-design-blocks~async-load-design-playground~async-load-gutenberg-blocks~~5a1097cf   -54393 B  (deleted)     -33695 B  (deleted)    -6961 B  (deleted)
async-load-design~async-load-design-blocks~async-load-design-playground~checkout~domains~purchases~w~ea89a787    +6502 B    (+3.8%)      +1609 B    (+1.2%)     +349 B    (+1.7%)
async-load-design~async-load-design-playground~async-load-gutenberg-blocks~async-load-gutenberg-comp~b2c3ace8       +0 B                    +0 B                  -1 B    (-0.0%)
async-load-design~async-load-design-playground~async-load-gutenberg-blocks~async-load-gutenberg-comp~d8c00363    +5534 B    (+3.8%)      +2903 B    (+3.1%)     +618 B    (+3.3%)
async-load-docs-selectors                                                                                           +0 B                    +0 B                  +1 B    (+0.1%)
async-load-featured-image                                                                                           +0 B                    +0 B                  +1 B    (+0.1%)
async-load-gutenberg-blocks                                                                                      +3622 B   (+11.4%)      +2615 B   (+13.4%)     +450 B    (+8.4%)
async-load-gutenberg-components                                                                                  +3622 B   (+10.2%)      +2615 B   (+12.1%)     +440 B    (+7.9%)
async-load-layout-guided-tours~checkout~plugins                                                                     +0 B                    +0 B                  -1 B    (-0.0%)
async-load-layout-masterbar-drafts-popover                                                                          +0 B                    +0 B                  -1 B    (-0.0%)
async-load-lib-abtest-test-helper                                                                                   +0 B                    +0 B                  +1 B    (+0.1%)
async-load-lib-happychat-connection                                                                                 +0 B                    +0 B                  -1 B    (-0.1%)
async-load-lib-network-connection                                                                                   +0 B                    +0 B                  -1 B    (-0.1%)
async-load-my-sites-current-site-domain-warnings                                                                    +0 B                    +0 B                  +2 B    (+0.1%)
async-load-my-sites-site-settings-section-export                                                                    +0 B                    +0 B                  -1 B    (-0.0%)
async-load-post-editor-editor-post-formats-accordion                                                                +0 B                    +0 B                  +1 B    (+0.1%)
async-load-post-editor-editor-seo-accordion                                                                         +0 B                    +1 B    (+0.0%)       +1 B    (+0.0%)
async-load-post-editor-editor-sharing-accordion                                                                     +0 B                    +0 B                  -1 B    (-0.0%)
async-load-post-editor-media-modal                                                                                  +0 B                    +0 B                  +1 B    (+0.0%)
async-load-post-editor-media-modal~media~post-editor                                                             +3306 B    (+1.5%)      +1038 B    (+1.0%)     +417 B    (+1.8%)
async-load-reader-feed-stream~async-load-reader-site-stream                                                      +7139 B   (+16.6%)      +1640 B    (+6.2%)     +504 B    (+8.8%)
async-load-reader-following-manage                                                                               +3187 B    (+4.5%)      +1262 B    (+3.6%)     +463 B    (+5.1%)
async-load-reader-following-manage~async-load-reader-search-stream                                               +7139 B   (+16.3%)      +1640 B    (+6.3%)     +508 B    (+8.6%)
async-load-reader-list-stream                                                                                       +0 B                    +0 B                  -1 B    (-0.0%)
async-load-reader-search-stream                                                                                     +0 B                    +0 B                  -1 B    (-0.0%)
async-load-reader-sidebar                                                                                       +17021 B   (+25.3%)      +5490 B   (+17.8%)    +1392 B   (+19.5%)
async-load-reader-tag-stream-main                                                                                +9882 B   (+33.1%)      +3850 B   (+26.4%)     +990 B   (+24.6%)
build                                                                                                          -401827 B   (-12.2%)    -147512 B   (-10.3%)   -30237 B    (-9.4%)
checklist                                                                                                           +0 B                    +0 B                  +1 B    (+0.0%)
checklist~stats                                                                                                  +2866 B    (+5.5%)      +1232 B    (+4.0%)     +328 B    (+4.5%)
checkout                                                                                                         +5423 B    (+1.4%)       +953 B    (+0.5%)     +315 B    (+0.7%)
checkout~customize~plugins~purchases~settings~theme~themes                                                       +1199 B    (+1.6%)       +640 B    (+2.5%)     +193 B    (+2.7%)
checkout~domains                                                                                                 +3622 B    (+5.7%)      +2615 B    (+7.4%)     +478 B    (+5.5%)
checkout~purchases                                                                                                  +0 B                    +0 B                  -1 B    (-0.0%)
comments                                                                                                        -21270 B    (-7.3%)     -12236 B    (-9.0%)    -5513 B   (-17.0%)
concierge                                                                                                       +26738 B   (+23.1%)      +9620 B   (+18.2%)    +2145 B   (+18.2%)
devdocs                                                                                                             +0 B                    +0 B                  +1 B    (+0.0%)
domains                                                                                                           +983 B    (+0.2%)       -345 B    (-0.1%)     +144 B    (+0.3%)
feature-upsell                                                                                                      +0 B                    +0 B                  -2 B    (-0.0%)
google-my-business                                                                                               +5528 B    (+4.0%)      +2284 B    (+3.4%)     +511 B    (+3.3%)
google-my-business~stats                                                                                         +5915 B   (+11.8%)      +2375 B    (+9.6%)     +364 B    (+6.7%)
help                                                                                                                +0 B                    +0 B                  +1 B    (+0.0%)
jetpack-connect                                                                                                  +8084 B    (+3.3%)      +3908 B    (+3.2%)     +827 B    (+3.0%)
jetpack-connect~plans~signup                                                                                     -4025 B    (-4.1%)      -2990 B    (-6.0%)     -384 B    (-3.1%)
jetpack-onboarding                                                                                              +10604 B    (+9.9%)      +4405 B    (+8.7%)    +1032 B    (+8.9%)
login                                                                                                             +732 B    (+0.4%)      -1422 B    (-1.7%)     -156 B    (-0.8%)
manifest                                                                                                            +0 B                  +882 B    (+2.1%)     +101 B    (+1.5%)
media                                                                                                            +6903 B    (+9.6%)      +2309 B    (+5.9%)     +522 B    (+5.7%)
notification-settings                                                                                            +2727 B    (+2.2%)       +772 B    (+1.2%)     +283 B    (+2.1%)
people                                                                                                           +1797 B    (+0.6%)      +1193 B    (+0.8%)     +222 B    (+0.7%)
plans                                                                                                           +18739 B   (+11.1%)     +10490 B   (+13.0%)    +2204 B   (+11.9%)
plugins                                                                                                         +14549 B    (+3.4%)      +7912 B    (+3.5%)    +1836 B    (+3.7%)
post-editor                                                                                                     +26801 B    (+2.4%)     +10427 B    (+2.1%)    +2447 B    (+1.9%)
post-editor~settings-writing~woocommerce                                                                         +1825 B    (+4.3%)      +1422 B    (+7.1%)     +242 B    (+4.4%)
posts-custom~posts-pages                                                                                         -2779 B    (-1.0%)      -2296 B    (-1.7%)     -407 B    (-1.4%)
posts-pages                                                                                                         +0 B                    +0 B                  +1 B    (+0.0%)
purchases                                                                                                        +1124 B    (+0.3%)       +624 B    (+0.4%)     +165 B    (+0.4%)
reader                                                                                                          +18581 B    (+8.7%)      +8909 B    (+8.1%)    +2389 B    (+9.6%)
security                                                                                                        +16197 B    (+6.8%)      +4147 B    (+3.3%)     +918 B    (+3.4%)
settings                                                                                                         -9325 B    (-3.0%)      -6869 B    (-4.3%)    -1503 B    (-3.9%)
settings-discussion                                                                                              -1797 B    (-2.7%)      -1193 B    (-3.3%)     -236 B    (-3.0%)
settings-security                                                                                               -11796 B   (-10.0%)      -8462 B   (-13.2%)    -1318 B    (-9.6%)
settings-traffic                                                                                                 -1797 B    (-0.8%)      -1193 B    (-1.2%)     -225 B    (-1.0%)
settings-writing                                                                                                 +5078 B    (+1.7%)       +887 B    (+0.6%)     +322 B    (+0.9%)
settings~settings-discussion~settings-security~settings-traffic~settings-writing                                +17053 B   (+30.9%)      +8158 B   (+26.6%)    +1687 B   (+25.1%)
sharing                                                                                                         +11443 B    (+4.9%)      +4659 B    (+4.0%)     +855 B    (+3.5%)
signup                                                                                                          -19896 B    (-4.4%)     -12621 B    (-5.0%)    -2619 B    (-4.5%)
stats                                                                                                            +6903 B    (+1.8%)      +2309 B    (+1.1%)     +578 B    (+1.3%)
theme~themes                                                                                                        +0 B                    +0 B                  +1 B    (+0.0%)
vendors~account~checkout~comments~domains~signup~stats                                                              +0 B                    +0 B                  +1 B    (+0.0%)
vendors~activity~async-load-blocks-calendar-popover~async-load-design~async-load-design-blocks~async~5c6393c9       +0 B                    +0 B                  +1 B    (+0.0%)
vendors~activity~async-load-design~async-load-design-blocks~async-load-design-playground~async-load-~ad6dafda       +0 B                    +0 B                  +1 B    (+0.0%)
vendors~async-load-components-web-preview-component                                                                 +0 B                    +0 B                  +1 B    (+0.0%)
vendors~async-load-design~async-load-design-blocks~async-load-design-playground~async-load-gutenberg~2306410c       +0 B                    +0 B                  +1 B    (+0.0%)
vendors~async-load-design~async-load-design-playground~async-load-gutenberg-blocks~async-load-gutenb~cc1a28e7       +0 B                    +0 B                  +1 B    (+0.0%)
vendors~async-load-lib-happychat-connection                                                                         +0 B                    +0 B                  +1 B    (+0.0%)
vendors~async-load-reader-following-manage~async-load-reader-search-stream                                          +0 B                    +0 B                  -1 B    (-0.0%)
vendors~devdocs                                                                                                     +0 B                    +0 B                  +1 B    (+0.0%)
vendors~post-editor~woocommerce                                                                                     +0 B                    +0 B                  -1 B    (-0.0%)
vendors~settings-writing                                                                                            +0 B                    +0 B                  -1 B    (-0.0%)
vendors~wp-super-cache                                                                                              +0 B                    +0 B                  +1 B    (+0.0%)
woocommerce                                                                                                     -22567 B    (-1.0%)     -12237 B    (-1.4%)    -2736 B    (-1.4%)
accept-invite~account~account-close~async-load-design-blocks~happychat~help~jetpack-connect~login~me~287a1d3e   +41400 B      (new)     +18895 B      (new)    +4613 B      (new)
account~async-load-design~async-load-design-blocks~async-load-design-playground~async-load-gutenberg~36c35619   +31091 B      (new)     +16158 B      (new)    +4479 B      (new)
async-load-components-web-preview-component~comments~reader                                                     +75811 B      (new)     +30883 B      (new)   +10106 B      (new)
async-load-design~async-load-design-playground~async-load-gutenberg-blocks~async-load-gutenberg-comp~a8299013   +51996 B      (new)     +31800 B      (new)    +6721 B      (new)
async-load-design~async-load-design-playground~async-load-gutenberg-blocks~async-load-gutenberg-comp~c7e83cb0   +38553 B      (new)     +33933 B      (new)   +10319 B      (new)
settings-security~signup                                                                                        +32118 B      (new)     +18912 B      (new)    +5242 B      (new)
```

dmsnell committed:
Data Layer: Load concierge data handlers dynamically

See #25073

This patch breaks apart the coupling in the data layer for the concierge
chats so that the handlers are only bundled into the application when
they are needed.

This is an ongoing effort to optimize and modularize our state system.

flootr commented on client/state/data-layer/wpcom/sites/index.js:
```diff
@@ -1,48 +0,0 @@
-		commentsTree,
-		config.isEnabled( 'jitms' ) ? jitm : null,
-		media,
-		config.isEnabled( 'memberships' ) ? memberships : null,
```
Only asking because I'm not sure about the impact of removing this. The config setting is `false` so it wasn't intended to be included. How is it now? Are those action files which now import these data layer handlers too only loaded if this config setting is `true`?

dmsnell replied:
the action files are conditionally loaded based on the `config` value - I checked. thanks for scrutinizing the code at this level though! it means you're doing a nice review!

flootr commented:
I tested the cycles and as far as I can see they shouldn't be a problem. Tested if the functionality of the relevant parts of Calypso still work as expected.

dmsnell commented:
@flootr thanks for testing - we may need to review a few things again because I rebased against `master` and squashed the commits (see the updated summary)

flootr commented:
/cc @adambbecker @michaeldcain

blowery commented:
@flootr @dmsnell what's left on this? I'd love to get that sweet 30k drop on build.

sirreal commented:
Deployment plans: p4TIVU-99U-p2

dmsnell commented:
@bluefuton it appears like `READER_SITE_BLOCKS_REQUEST` was never being dispatched and so after this change we had no action-creator from which to import the state module. is it right in your eyes that this data layer handler should never load inside of Calypso?

@artpi you might want to double-verify that these changes work for the mailchimp integration since that was added after this PR was created and we had to rebase it in.

adamziel commented:
@dmsnell just tested the manual merge of #27245 and it works well, thank you!

bluefuton commented:
> @bluefuton it appears like READER_SITE_BLOCKS_REQUEST was never being dispatched and so after this change we had no action-creator from which to import the state module. is it right in your eyes that this data layer handler should never load inside of Calypso?

@dmsnell well spotted! The site block listing is in the process of being built. The data layer bits have merged, and the front end is being worked on in https://github.com/Automattic/wp-calypso/pull/27333, so we're not dispatching `READER_SITE_BLOCKS_REQUEST` in master just yet.
